src/types.ts

```diff
@@ -0,0 +1,154 @@
+  columnDeclaredType: string;
+  /**
+   * The index for this column for this resultset*/
+  columnIndex: number;
+};
+
+/**
+ * Allows the execution of bulk of sql commands
+ * inside a transaction
+ * If a single query must be executed many times with different arguments, its preferred
+ * to declare it a single time, and use an array of array parameters.
+ */
+export type SQLBatchTuple = [string] | [string, Array<any> | Array<Array<any>>];
+
+/**
+ * status: 0 or undefined for correct execution, 1 for error
+ * message: if status === 1, here you will find error description
+ * rowsAffected: Number of affected rows if status == 0
+ */
+export type BatchQueryResult = {
+  rowsAffected?: number;
+};
+
+/**
+ * Result of loading a file and executing every line as a SQL command
+ * Similar to BatchQueryResult
+ */
+export interface FileLoadResult extends BatchQueryResult {
+  commands?: number;
+}
+
+export enum RowUpdateType {
+  SQLITE_INSERT = 18,
+  SQLITE_DELETE = 9,
+  SQLITE_UPDATE = 23
+}
+export interface UpdateNotification {
+  opType: RowUpdateType;
+  table: string;
+  rowId: number;
+}
+
+export type UpdateCallback = (update: UpdateNotification) => void;
+
+export type ContextLockID = string;
+
+export enum ConcurrentLockType {
+  READ,
+  WRITE
+}
+
+export type OpenOptions = {
+  location?: string;
+  /**
+   * The number of concurrent read connections to use.
+   * Setting this value to zero will only open a single write connection.
+   * Setting this value > 0 will open the DB in WAL mode with [numReadConnections]
+   * read connections and a single write connection. This allows for concurrent
+   * read operations during a write operation.
+   */
+  numReadConnections?: number;
+};
+
+export type Open = (dbName: string, options?: OpenOptions) => QuickSQLiteConnection;
+
+export interface ISQLite {
+  open: Open;
+  close: (dbName: string) => void;
+  delete: (dbName: string, location?: string) => void;
+
+  requestLock: (dbName: string, id: ContextLockID, type: ConcurrentLockType) => QueryResult;
+  releaseLock(dbName: string, id: ContextLockID): void;
+  executeInContext: (dbName: string, id: ContextLockID, query: string, params: any[]) => QueryResult;
+  executeInContextAsync: (dbName: string, id: ContextLockID, query: string, params: any[]) => Promise<QueryResult>;
+
+  attach: (mainDbName: string, dbNameToAttach: string, alias: string, location?: string) => void;
+  detach: (mainDbName: string, alias: string) => void;
+
+  executeBatchAsync: (dbName: string, commands: SQLBatchTuple[], id: ContextLockID) => Promise<BatchQueryResult>;
+  loadFileAsync: (dbName: string, location: string) => Promise<FileLoadResult>;
+}
+
+export interface LockOptions {
+  timeoutMs?: number;
+}
+
+export interface LockContext {
+  execute: (sql: string, args?: any[]) => QueryResult;
+  executeAsync: (sql: string, args?: any[]) => Promise<QueryResult>;
```

Comment:
It may be better to just have a single asynchronous `execute` method here, same as on the connection.

Reply:
👌 do you think the transaction contexts should only have asynchronous `commit` and `rollback` operations as well?

Reply:
Yes, I think it's best if it's all asynchronous. That leaves the option open to change the implementation later, and then same API could also be implemented on other platforms that can only do asynchronous calls.

Reply:
👍 The lock and transaction methods are now all async